### ISSUE TITLE
403 Forbidden when connecting 10+ orgs

### ISSUE THREAD

**Issue by SerKnight** [XeroAPI/xero-node#428]
**SDK you're using (please complete the following information):**
 - 4.6.0

**Describe the bug**
`OPError: expected 200 OK, got: 403 Forbidden`

**To Reproduce**
* connect 10+ orgs

**Expected behavior**
* successful API responses and token_sets

**Screenshots**
```
OPError: expected 200 OK, got: 403 Forbidden
        at processResponse (/srv/www/magiqsandbox/releases/release-20200623221229/node_modules/openid-client/lib/helpers/process_response.js:48:11)
        at /srv/www/magiqsandbox/releases/release-20200623221229/node_modules/openid-client/lib/issuer.js:247:20
        at async /srv/www/magiqsandbox/releases/release-20200623221229/node_modules/p-some/index.js:53:19
    OPError: expected 200 OK, got: 403 Forbidden
        at processResponse (/srv/www/magiqsandbox/releases/release-20200623221229/node_modules/openid-client/lib/helpers/process_response.js:48:11)
        at /srv/www/magiqsandbox/releases/release-20200623221229/node_modules/openid-client/lib/issuer.js:247:20
```

**Additional context**
* happening for a migrated Partner app and their new OAuth2 connections. ( these are not migrated tokens, but new in  OAuth2)

**Comment by SerKnight** [comment]
If anyone else has encountered this please report/].

**Comment by TheWass** [comment]
@SerKnight, We have found a solution to this.  
Our past code had this library instantiated once per org.  This made sense during oauth1 where each connection could only support one org.  Now that oauth2 supports many orgs, our code still instantiated the library once per org.  When a user connects with many organizations, that means many instances were created and asked for openid information at the same time.

We solved this by building a manager to share the library among all orgs for the entire connection.  

In summary, for other people who may run into this issue, make sure you aren't instantiating xero-node multiple times at the same time.

**Comment by RettBehrens** [comment]
Hey @TheWass thanks for sharing an update on this! Going to close this for now but if you or another dev encounters this issue again feel free to reopen.

**Comment by brandonn1231** [comment]
@RettBehrens @TheWass sorry, has this been fixed in all releases of xero-node? I'm currently running v4.7.6. I'm currently running into this 403 Forbidden error, but I'm not sure if its because of the same issue as in this thread. I made another post here https://github.com/XeroAPI/xero-node/issues/486

**Comment by TheWass** [comment]
@brandonn1231, my "Fix" is more of a redesign of the use of the library, rather than a fix to the library itself.  The problem happened during the `initialize()` call.  For my case, multiple calls were happening at the same time.  So, the solution is to reduce redundant calls.

To elaborate a bit more, I've created a class to manage the connections by user.  So, instead of looping through all of the organizations, and creating xero-node instances for each org, I create a singleton xero-node instance for the user, and store it in a Map class.  

Incomplete example showing the main beats:
```
private static instances: Map<string, ClientManager>
public xeroClient: XeroClient;
private constructor() { }
public static async getInstance(id: string) {
    let instance = ClientManager.instances.get(id);
    if (!instance) {
        instance = new ClientManager();
        instance.xeroClient = await (new XeroClient(config)).initialize()
        ClientManager.instances.set(id, instance);
    }
    return instance;
}
public static destroy(id: string) {
    ClientManager.instances.delete(id);
}
```
Yes, this pattern is based on the singleton pattern.  But it's an efficient mechanism to reduce the number of calls to `initialize()`.  It does require a bit more management to ensure the instances are deleted so you don't get memory leaks.
